lib/ui/compositing.dart

```diff
@@ -387,13 +387,35 @@ class SceneBuilder extends NativeFieldWrapperClass2 {
   /// {@macro dart.ui.sceneBuilder.oldLayerVsRetained}
   ///
   /// See [pop] for details about the operation stack.
-  ColorFilterEngineLayer pushColorFilter(Color color, BlendMode blendMode, { ColorFilterEngineLayer oldLayer }) {
+  ColorFilterEngineLayer pushColorFilter(ColorFilter filter, { ColorFilterEngineLayer oldLayer }) {
+    assert(filter != null);
+    assert(filter._type != null && filter._type != ColorFilter._TypeNone);
     assert(_debugCheckCanBeUsedAsOldLayer(oldLayer, 'pushColorFilter'));
-    final ColorFilterEngineLayer layer = ColorFilterEngineLayer._(_pushColorFilter(color.value, blendMode.index));
+    EngineLayer engineLayer;
+    switch(filter._type) {
+      case ColorFilter._TypeMode:
+        engineLayer = _pushColorFilter(filter._color.value, filter._blendMode.index);
+        break;
+      case ColorFilter._TypeMatrix:
+        engineLayer = _pushColorFilterMatrix(Float32List.fromList(filter._matrix));
+        break;
+      case ColorFilter._TypeLinearToSrgbGamma:
+        engineLayer = _pushColorFilterLinearToSrgbGamma();
+        break;
+      case ColorFilter._TypeSrgbToLinearGamma:
+        engineLayer = _pushColorFilterSrgbToLinearGamma();
+        break;
+      default:
+        assert(false);
+    }
+    final ColorFilterEngineLayer layer = ColorFilterEngineLayer._(engineLayer);
     assert(_debugPushLayer(layer));
     return layer;
   }
   EngineLayer _pushColorFilter(int color, int blendMode) native 'SceneBuilder_pushColorFilter';
```

Comment:
Instead of implementing 4 pushColorFilter variants, make it's cleaner to define a single `pushColorFilter` that takes a `sk_sp<SkColorFilter>`, and expose `ExtractColorFilter` from `paint.cc` to generate the `SkColorFilter`.

Reply:
My concern with that is the Pairthent version is a bit hacky and was mainly done to maintain backwards compatibility.

Paint can afford to do this because it already encodes/serializes this data to the engine side.  Doing that is not free.  I'd rather use this approach because it avoids allocating unnecessary typed data (which is demonstrably slow), and avoids sending data over to thenative side that we don't need.

Reply:
Thanks for pointing that out! I didn't realize that `ColorFilter` doesn't have a corresponding native `SkColorFilter` object. Is it possible to create such a native object during the construction of `ColorFilter`, so `Paint` can just store it inside `_objects` (like how `Shader` is stored), and `pushColorFilter` can also directly use it without triggering additional Dart to native conversion costs?

I guess that I'm Ok with having 4 `pushColorFilter***` in `scene_builder.cc` but it just feels a little weird as every previous `push***` in `scene_builder.cc` has a one-to-one relationship with the `SceneBuilder.push***` in `compositing.dart`. I'll consult @chinmaygarde and @yjbanov to see how they feel about 4 `pushColorFilter***` variants in `scene_builder.cc`.

Reply:
That may be a better way to go - just make ColorFilter a `NativeFieldWrapper2` class.  Do you want that to be part of this pull request or should we split it out?

Reply:
Let's have it here if it's not complicating this PR too much :) (I was hoping that it could even simplify this PR a little bit.)

Reply:
I split that effort out into #9668 which has landed and been merged into this PR.  PTAL.